src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java

```diff
@@ -64,63 +64,59 @@ public String handleNotifications(final String input) {
-                                item.getMerchantReference(),
-                                item.getMerchantAccountCode(),
-                                e);
-                }
-            }
-            response = "[accepted]";
-        }
-
-        try {
-            return createSendNotificationResponse(response);
-        } catch (final Exception e) {
-            // Not much we can do. Let Adyen retry.
-            throw new RuntimeException(e);
+        final List<NotificationRequestItem> listOfNotifications = sendNotification.getNotification()
+                                                                                  .getNotificationItems()
+                                                                                  .getNotificationRequestItem();
+        for (final NotificationRequestItem item : listOfNotifications) {
+            handleNotification(item);
         }
+        return createSendNotificationResponse("[accepted]");
     }
 
     private void handleNotification(final NotificationRequestItem item) {
-        logger.info("Handling notification: eventCode='{}', pspReference='{}', originalReference='{}', success='{}', reason='{}', merchantReference='{}', merchantAccount='{}'",
+        Exception error = null;
+        final long startTime = System.currentTimeMillis();
+        long duration = 0L;
+        try {
+            final AdyenNotificationHandler adyenNotificationHandler = getAdyenNotificationHandler(item);
+            if (adyenNotificationHandler == null) {
+                logger.warn("No handler available - ignoring");
+                return;
+            }
+
+            adyenNotificationHandler.handleNotification(item);
+            duration = System.currentTimeMillis() - startTime;
+        } catch (final Exception e) {
+            duration = System.currentTimeMillis() - startTime;
+            error = e;
+        } finally {
+            final String message = String.format(
+                    "op='notificationHandling' eventCode='%s', pspReference='%s', originalReference='%s', success='%s', reason='%s', merchantReference='%s', merchantAccount='%s', duration=%d, error=%s",
                     item.getEventCode(),
                     item.getPspReference(),
                     item.getOriginalReference(),
                     item.isSuccess(),
                     item.getReason(),
                     item.getMerchantReference(),
-                    item.getMerchantAccountCode());
-
-        final AdyenNotificationHandler adyenNotificationHandler = getAdyenNotificationHandler(item);
-        if (adyenNotificationHandler == null) {
-            logger.warn("No handler available - ignoring");
-            return;
+                    item.getMerchantAccountCode(),
+                    duration,
+                    error != null);
+
+            if (error == null) {
+                logger.info(message);
+            } else {
+                logger.error(message, error);
```

Comment:
I was thinking of using the error level for all error scenarios (which reminds me I need to fix the `logTransactionError` method), but I don't really have a strong opinion about it.

Is KB core using warn for exception handling?